pkg/db/pebble.go

```diff
@@ -0,0 +1,266 @@
+	var start time.Time
+	option.EventListener.WriteStallBegin = func(_ pebble.WriteStallBeginInfo) {
+		atomic.AddInt64(&ws.counter, 1)
+		start = time.Now()
+	}
+	option.EventListener.WriteStallEnd = func() {
+		duration := time.Since(start)
+		ws.duration.Store(duration)
+	}
+	return option, ws
+}
+
+// OpenPebble opens a pebble.
+func OpenPebble(ctx context.Context, id int, path string, cfg *config.DBConfig) (DB, error) {
+	option, ws := buildPebbleOption(id, cfg)
+	dbDir := filepath.Join(path, fmt.Sprintf("%04d", id))
+	err := retry.Do(ctx, func() error {
+		err1 := os.RemoveAll(dbDir)
+		if err1 != nil {
+			log.Warn("clean data dir fails",
+				zap.String("dir", dbDir), zap.Error(err1))
+		}
+		return err1
+	},
+		retry.WithBackoffBaseDelay(500), // 0.5s
+		retry.WithMaxTries(4))           // 2s in total.
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	db, err := pebble.Open(dbDir, &option)
+	if err != nil {
+		return nil, err
+	}
+	return &pebbleDB{
+		db:               db,
+		metricWriteStall: ws,
+	}, nil
+}
+
+// Metrics of pebble write stall.
+type writeStall struct {
+	counter  int64
+	duration atomic.Value
+}
+
+type pebbleDB struct {
+	db               *pebble.DB
+	metricWriteStall *writeStall
+}
+
+var _ DB = (*pebbleDB)(nil)
+
+func (p *pebbleDB) Snapshot() (Snapshot, error) {
+	return pebbleSnapshot{Snapshot: p.db.NewSnapshot()}, nil
+}
+
+func (p *pebbleDB) Batch(cap int) Batch {
+	return pebbleBatch{
+		Batch: p.db.NewBatch(),
+	}
+}
+
+func (p *pebbleDB) Close() error {
+	return p.db.Close()
+}
+
+// TODO: Update metrics once we switch to pebble,
+//       as some metrics are not applicable to pebble.
+func (p *pebbleDB) CollectMetrics(captureAddr string, i int) {
+	db := p.db
+	stats := db.Metrics()
+	id := strconv.Itoa(i)
+	sum := 0
+	for i := range stats.Levels {
+		sum += int(stats.Levels[i].Size)
+	}
+	sorter.OnDiskDataSizeGauge.
+		WithLabelValues(captureAddr, id).Set(float64(stats.DiskSpaceUsage()))
+	sorter.InMemoryDataSizeGauge.
+		WithLabelValues(captureAddr, id).Set(float64(stats.BlockCache.Size))
+	dbIteratorGauge.
+		WithLabelValues(captureAddr, id).Set(float64(stats.TableIters))
+	dbWriteDelayCount.
+		WithLabelValues(captureAddr, id).
+		Set(float64(atomic.LoadInt64(&p.metricWriteStall.counter)))
+	stallDuration := p.metricWriteStall.duration.Load()
+	if stallDuration != nil {
+		p.metricWriteStall.duration.Store(nil)
+		dbWriteDelayDuration.
+			WithLabelValues(captureAddr, id).
+			Set(stallDuration.(time.Duration).Seconds())
+	}
+	metricLevelCount := dbLevelCount.
+		MustCurryWith(map[string]string{"capture": captureAddr, "id": id})
+	for level, metric := range stats.Levels {
+		metricLevelCount.WithLabelValues(fmt.Sprint(level)).Set(float64(metric.NumFiles))
+	}
+}
+
+type pebbleBatch struct {
+	*pebble.Batch
+}
+
+var _ Batch = (*pebbleBatch)(nil)
+
+func (b pebbleBatch) Put(key, value []byte) {
+	_ = b.Batch.Set(key, value, pebble.NoSync)
+}
+
+func (b pebbleBatch) Delete(key []byte) {
```

Comment:
It's a bit strange that lint should be complaining that you didn't write a comment for a public method, right?

Reply:
Yeah, it may due to the private receiver `pebbleBatch`.